docs/blog/rustport/index.html

```diff
@@ -0,0 +1,354 @@
+
+<h2 id="the-story-of-the-port">The Story Of The Port</h2>
+
+<p>We had decided we were gonna do a “Fish <a href="https://en.wikipedia.org/wiki/Ship_of_Theseus">Of Theseus</a>” port - we would move over, component by component, until no C++ was left.
+And at every stage of that process, it would remain a working fish.</p>
+
+<p>This was a necessity - if we didn’t, we would not have a working program for months, which is not only demoralizing but would also have precluded us from
+using most of our test suite - which is end-to-end tests that run a script or fake a terminal interaction. We would also not have been able to do another C++ release,
+putting some cool improvements into the hands of our users.</p>
+
+<p>Had we chosen to disappear into a hole we might not have finished at all, and we would have to re-do a bunch of work once it became testable.
+We also mostly kept the structure of the C++ code intact - if a function is in the “env” subsystem, it would stay there. Resisting the temptation to
+clean up allowed us to compare the before and after to find places where we had mistranslated something.</p>
+
+<p>So we used <a href="https://google.github.io/autocxx/">autocxx</a> to generate bindings between C++ and Rust code, allowing us to port one component at a time.</p>
+
+<p>We started<sup id="fnref:technically"><a href="#fn:technically" class="footnote" rel="footnote" role="doc-noteref">3</a></sup> by porting the builtins. These are essentially little self-contained programs, with their own arguments, streams, exit code, etc.
+That means it’s easy to port them separately from the rest of the shell once you have a way to call a Rust builtin from C++, which we had as part of the initial pull request.</p>
+
+<p>Where they connected to the main shell, we used one of three approaches:</p>
+
+<ol>
+  <li>Add some FFI glue to the C++ to make it callable from Rust, port the caller and leave the callee for later</li>
+  <li>Move the callee to Rust and, if necessary, make it callable from C++</li>
+  <li>Write a Rust version of the callee and call it from the ported caller, but leave the C++ version around</li>
+</ol>
+
+<p>For instance, almost every builtin needs to parse its options. We have our own implementation of getopt, that we reimplemented in Rust in the initial PR,
+but the C++ version stuck around until it had no more callers remaining. Otherwise we would have had to write a C++-to-Rust bridge and adjust the C++ callers to use it.</p>
+
+<p>Or the <code class="language-plaintext highlighter-rouge">builtin</code> builtin needs access to the names of all builtins to print them for <code class="language-plaintext highlighter-rouge">builtin --get-names</code>. In that case we bridged some access to what amounts to a constant vector of strings in the C++, and eventually moved it over once the users were in Rust.</p>
```

Comment:
Do you want this `builtin builtin` repetition right ?

```suggestion
<p>Or the <code class="language-plaintext highlighter-rouge">builtin</code> needs access to the names of all builtins to print them for <code class="language-plaintext highlighter-rouge">builtin --get-names</code>. In that case we bridged some access to what amounts to a constant vector of strings in the C++, and eventually moved it over once the users were in Rust.</p>
```

Reply:
Yes, it's about the builtin called "builtin".

Reply:
Thanks for the others, I've added a comment here.